src/main/java/de/qabel/core/module/ModuleManager.java

```
@@ -87,50 +75,57 @@ public void run() {
 	 * @throws InstantiationException
 	 * @throws IllegalAccessException
 	 */
-	public void startModule(Class<?> module) throws InstantiationException, IllegalAccessException {
-		Module m = (Module) module.newInstance();
-		m.setModuleManager(this);
+	public <T extends Module> T startModule(Class<T> module) throws InstantiationException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
+		T m = module.getConstructor(ModuleManager.class).newInstance(this);
 		m.init();
         ModuleThread t = new ModuleThread(m);
-		getModules().put(m, t);
+		modules.put(m, t);
 		t.start();
+		return m;
 	}
-	
+
 	public void startModule(File jar, String className) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
 		try {
 			ClassLoader cld = LOADER;
 			cld.addURL(jar.toURI().toURL());
-			startModule(Class.forName(className, true, cld));
-		} catch (MalformedURLException e) {
+			Class cls = Class.forName(className, true, cld);
+			try {
+				startModule(cls.asSubclass(Module.class));
+			}
+			catch (ClassCastException e) {
+				throw new InstantiationError(className + " not a subclass of Module!");
+			}
+		} catch (Exception e) {
 			e.printStackTrace();
-			// Should not happen!
 			System.exit(1);
```

Do we really want to exit here? I'd throw a runtime exception.
(I know you didn't write this, but at least you removed the line comment ;-))